packages/eas-cli/src/rollout/branch-mapping.ts

```diff
@@ -0,0 +1,265 @@
+import assert from 'assert';
+
+import {
+  BranchMapping,
+  BranchMappingNode,
+  assertNodeObject,
+  assertNumber,
+  assertStatement,
+  assertString,
+  getBranchMapping,
+  getNodesFromStatement,
+  isAlwaysTrue,
+  isAndStatement,
+  isStatement,
+} from '../channel/branch-mapping';
+import { getUpdateBranch } from '../channel/utils';
+import { UpdateChannelBasicInfoFragment } from '../graphql/generated';
+import { UpdateBranchObject, UpdateChannelObject } from '../graphql/queries/ChannelQuery';
+
+export type Rollout = LegacyRollout | ConstrainedRollout;
+export type RolloutInfo = LegacyRolloutInfo | ConstrainedRolloutInfo;
+type ConstrainedRollout = LegacyRollout & {
+  runtimeVersion: string;
+};
+
+type LegacyRollout = {
+  rolledOutBranch: UpdateBranchObject;
+  defaultBranch: UpdateBranchObject;
+} & LegacyRolloutInfo;
+
+type ConstrainedRolloutInfo = LegacyRolloutInfo & {
+  runtimeVersion: string;
+};
+
+type LegacyRolloutInfo = {
+  rolledOutBranchId: string;
+  percentRolledOut: number;
+  defaultBranchId: string;
+};
+
+export function isLegacyRolloutInfo(rollout: RolloutInfo): rollout is LegacyRolloutInfo {
+  return !isConstrainedRolloutInfo(rollout);
+}
+
+export function isConstrainedRolloutInfo(rollout: RolloutInfo): rollout is ConstrainedRolloutInfo {
+  return 'runtimeVersion' in rollout;
+}
+
+export function isConstrainedRollout(rollout: Rollout): rollout is ConstrainedRollout {
+  return isConstrainedRolloutInfo(rollout);
+}
+
+export function getRolloutInfo(basicChannelInfo: UpdateChannelBasicInfoFragment): RolloutInfo {
+  const branchMapping = getBranchMapping(basicChannelInfo.branchMapping);
+  assertRollout(branchMapping);
```

Comment:
it might be better to either a) combine these (so that `getBranchMapping` never returns an invalid branch mapping) or b) rename `getBranchMapping` to `parseBranchMappingJSON` and have it return an `object` rather than a `BranchMapping`.

Reply:
so the issue here is that all rollouts are branch mappings but not all branch mappings are rollouts. `assertRollout` ensures the branch mapping has the rolloutToken among other things, as opposed to the standard 'alwaysTrue' `channel->branch` mapping.